Observability namespace

aabeshov opened the pull request:
### Description

Adding observability namespace specs.

### Issues Resolved

Closes #230.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock reviewed (changes requested):
Great start! Needs tests. 

You can add these specs piecemeal instead of everything at once, maybe start with something simple with tests, changelog, etc.?

nhtruong commented:
Agree with @dblock. This is a lot in one PR esp once you add the tests. Do break it into smaller PRs and thank you so much for the hard work :)

aabeshov commented:
> Agree with @dblock. This is a lot in one PR esp once you add the tests. Do break it into smaller PRs and thank you so much for the hard work :)

It's okay ,thanks, I guess I am close to the finish line with these specs
Check Links is failing, is it connected with test stories that I create ?
Because my test stories are passing locally
<img width="922" alt="image" src="https://github.com/user-attachments/assets/4bc77979-73da-45ea-bd6a-94b32f70571d">

dblock commented:
Looks like only the link checker is failing against `http://localhost:9090` which is mentioned in a test. 

Add an exclude file like [here](https://github.com/opensearch-project/OpenSearch/blob/main/.github/workflows/links.yml#L18C90-L18C121) or change the test?

dblock commented on tests/observability/observability.yaml:
```diff
@@ -0,0 +1,162 @@
+        totalHits: 0
+        totalHitRelation: eq
+        observabilityObjectList: []
+
```
Nit: let's get rid of double empty lines ;)

aabeshov commented:
I changed the lychee ignore. --exclude-file is deprecated.
BTW, I haven't implemented  /_plugins/_query/settings: ["PUT"] this api , because I guess @Tokesh has already made it in this PR [#456 ]

dblock commented on .cspell:
```diff
@@ -185,4 +185,6 @@ urldecode
-Yrtsd
\ No newline at end of file
+Yrtsd
+Refn
```
We actually removed this guy.

dblock commented on .lycheeignore:
```diff
@@ -1 +1,3 @@
 https://localhost:9200*
+http://localhost:9090
+http://localhost:9091
```
Maybe `https://localhost:*` ?

dblock commented on spec/namespaces/observability.yaml:
```diff
@@ -0,0 +1,224 @@
+            $ref: '../schemas/observability._common.yaml#/components/schemas/ObservabilityObject'
+  responses:
+    observability.list_objects@200:
+      description: Successful response of retrieving all Observability Objects
```
Add a period. (Note to self, we need a validator.)

dblock commented on tests/observability/datasources.yaml:
```diff
@@ -0,0 +1,94 @@
+$schema: ../../json_schemas/test_story.schema.yaml
```
This is for `query`, so it should live in `tests/query/datasources.yml`, right?

aabeshov replied:
I will move datasource test story to /tests/default/query/. But do you think I should move observability to default ?  Cause as far as I know observability is an external service

dblock replied:
This is a bit new in this repo.

Folders like `default` and `snapshot` are there to have separate docker-compose.yml, not because something is external or internal. These are "test suites". So you should move into `tests/default/query` and `tests/default/observability`. 

If observability required some fancy setup, you could create a new folder for it and add a job to the GHA, but it is hopefully not necessary.

dblock reviewed (changes requested):
This looks good! I have some small asks and let's get this merged.

aabeshov commented:
I don't know why but there is a bad request when test story is making DELETE /_plugins/_query/_datasources/{datasource_name}

dblock commented:
> @dblock ,I don't know why but there is a 400 bad request when test story is making DELETE /_plugins/_query/_datasources/{datasource_name} However locally it's okay

We should show the detailed output for these errors (please open an issue). 

For now, add a `--verbose` into the GHA workflow to see the details?

aabeshov commented:
> > @dblock ,I don't know why but there is a 400 bad request when test story is making DELETE /_plugins/_query/_datasources/{datasource_name} However locally it's okay
> 
> We should show the detailed output for these errors (please open an issue).
> 
> For now, add a `--verbose` into the GHA workflow to see the details? Note that it fails on older versions, maybe a quick start of a container with 1.3.x locally will reproduce it?

[INFO] => DELETE /_plugins/_query/_datasources/test_datasource ({}) [application/json] 
[INFO] <= 400 (application/json) | "no handler found for uri [/_plugins/_query/_datasources/test_datasource] and method [DELETE]"
[INFO] => DELETE /_plugins/_query/_datasources/test_datasource ({}) [application/json] 
[INFO] <= 400 (application/json) | "no handler found for uri [/_plugins/_query/_datasources/test_datasource] and method [DELETE]"

ERROR   query/datasources.yaml (/home/runner/work/opensearch-api-specification/opensearch-api-specification/tests/default/query/datasources.yaml)
    ERROR   PROLOGUES
        ERROR   DELETE /_plugins/_query/_datasources/{datasource_name} (Bad Request)
    SKIPPED CHAPTERS
        SKIPPED Create a new query datasource.
        SKIPPED Retrieve the list of all query datasources.
        SKIPPED Retrieve a specific query datasource by name.
        SKIPPED Update an existing query datasource.
        SKIPPED Retrieve the updated query datasource by name.
        SKIPPED Delete the query datasource by name.
    ERROR   EPILOGUES
        ERROR   DELETE /_plugins/_query/_datasources/{datasource_name} (Bad Request)
        
        Here is the details. I am not sure, but maybe there is not such api in old versions

Tokesh commented on spec/namespaces/observability.yaml:
```diff
@@ -0,0 +1,224 @@
+    get:
+      operationId: observability.get_localstats.0
+      x-operation-group: observability.get_localstats
+      x-version-added: '1.1'
```
please recheck `x-version-added`. i think it can fix your ci issue

aabeshov replied:
@Tokesh committed changes , you can check

dblock commented on .github/workflows/test-spec.yml:
```diff
@@ -76,6 +76,7 @@ jobs:
             --opensearch-insecure \
             --opensearch-version=${{ matrix.entry.version }} \
             --coverage coverage/test-spec-coverage-${{ matrix.entry.version }}-${{ matrix.entry.tests || 'default' }}.json \
+            --verbose \
```
Let's not leave this here, it spews a lot of logs. Remove.

dblock commented on .lycheeignore:
```diff
@@ -1 +1,2 @@
 https://localhost:9200*
+http://localhost:*
```
Let's consolidate this into a single `https://localhost:*` and change the prometheus URL below to an `https://` one so it catches both?

dblock commented on tests/default/query/datasources.yaml:
```diff
@@ -0,0 +1,95 @@
+        connector: PROMETHEUS
+        allowedRoles: []
+        properties:
+          prometheus.uri: 'http://localhost:9090'
```
This (and a few other instances below) is the URL that gets tripped by the link checker. Change it to `https://...`

I am surprised the URL is actually not checked, is this expected? I'd think the API fails if prometheus is not available?

aabeshov replied:
It sound like it should fail, but I tested API in the playground with turned off prometheus and got 200 without errors

dblock reviewed (changes requested):
Looks good. I have 2 asks below.

dblock commented:
Looks like one of the tests here may fail intermittently:

```
        FAILED  Retrieve list of Observability objects.
            PASSED  REQUEST BODY
            PASSED  RESPONSE STATUS
            FAILED  RESPONSE PAYLOAD BODY (expected totalHits='0', got '1')
            PASSED  RESPONSE PAYLOAD SCHEMA
            SKIPPED RESPONSE OUTPUT VALUES
```

https://github.com/opensearch-project/opensearch-api-specification/actions/runs/10354531782/job/28660019203?pr=493

@aabeshov care to take a look please?

dblock commented:
Thanks! I fixed it in https://github.com/opensearch-project/opensearch-api-specification/pull/497 since it was failing often. Does that look right? Next time don't hold back and just make a PR :)

aabeshov commented:
> Thanks! I fixed it in #497 since it was failing often. Does that look right? Next time don't hold back and just make a PR :)



> Thanks! I fixed it in #497 since it was failing often. Does that look right? Next time don't hold back and just make a PR :)

Thanks , but I am really confused because I chapters first I do get request to specific object and it return that It exists then right after this step I do request to get list of all objects and sometimes it returns that list is empty that why it is expected for index to be 0 in some cases.

dblock commented:
Writing something doesn't guarantee that it will be read right away, it's eventually consistent. Since these options are backed by an index, refreshing means waiting for a full refresh cycle where everything written gets read consistently, making it predictable. Does this help?

https://opensearch.org/docs/latest/api-reference/index-apis/flush/

aabeshov commented:
> Writing something doesn't guarantee that it will be read right away, it's eventually consistent. Since these options are backed by an index, refreshing means waiting for a full refresh cycle where everything written gets read consistently, making it predictable. Does this help?
> 
> https://opensearch.org/docs/latest/api-reference/index-apis/refresh/

aaa okay now I got this, thanks!
